security/src/main/scala/pl/touk/nussknacker/ui/security/api/AuthManager.scala

```diff
@@ -0,0 +1,173 @@
+package pl.touk.nussknacker.ui.security.api
+
+import akka.http.scaladsl.model.StatusCodes
+import akka.http.scaladsl.server.{
+  AuthorizationFailedRejection,
+  Directive0,
+  Directive1,
+  Directives,
+  RejectionHandler,
+  Route
+}
+import akka.http.scaladsl.server.Directives.{complete, handleRejections, optionalHeaderValueByName, provide, reject}
+import pl.touk.nussknacker.security.AuthCredentials.{
+  ImpersonatedAuthCredentials,
+  NoCredentialsProvided,
+  PassedAuthCredentials
+}
+import pl.touk.nussknacker.security.{AuthCredentials, ImpersonatedUserIdentity}
+import pl.touk.nussknacker.ui.security.api.AuthManager.{ImpersonationConsideringInputEndpoint, impersonateHeaderName}
+import pl.touk.nussknacker.ui.security.api.CreationError.ImpersonationNotAllowed
+import pl.touk.nussknacker.ui.security.api.SecurityError.{
+  CannotAuthenticateUser,
+  ImpersonatedUserDataNotFoundError,
+  ImpersonationMissingPermissionError,
+  InsufficientPermission
+}
+import sttp.tapir._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class AuthManager(protected val authenticationResources: AuthenticationResources)(implicit ec: ExecutionContext)
+    extends AkkaBasedAuthManager {
+
+  protected lazy val isAdminImpersonationPossible: Boolean =
+    authenticationResources.configuration.isAdminImpersonationPossible
+  protected lazy val authenticationRules: List[AuthenticationConfiguration.ConfigRule] =
+    authenticationResources.configuration.rules
+
+  def authenticate(authCredentials: AuthCredentials): Future[Either[AuthenticationError, AuthenticatedUser]] =
+    authCredentials match {
+      case NoCredentialsProvided => authenticateWithAnonymousAccess()
+      case passedCredentials @ PassedAuthCredentials(_) =>
+        authenticationResources.authenticate(passedCredentials).map(_.toRight(CannotAuthenticateUser))
+      case ImpersonatedAuthCredentials(impersonatingUserAuthCredentials, impersonatedUserIdentity) =>
+        authenticateWithImpersonation(impersonatingUserAuthCredentials, impersonatedUserIdentity)
+    }
+
+  def authenticationEndpointInput(): EndpointInput[AuthCredentials] =
+    authenticationResources
+      .authenticationMethod()
+      .withPossibleImpersonation()
+
+  def authorize(user: AuthenticatedUser): Either[AuthorizationError, LoggedUser] = {
+    if (user.roles.nonEmpty)
+      // TODO: This is strange that we call authenticator.authenticate and the first thing that we do with the returned user is
+      //       creation of another user representation based on authenticator.configuration. Shouldn't we just return the LoggedUser?
+      LoggedUser(user, authenticationRules, isAdminImpersonationPossible) match {
```

Comment:
1. By convention, companion's object apply method rather returns companion class instance. For method returning either we can create a named factory method e.g. `LoggedUser.create(...)`.
2. It can be simplified to `.left.map(...)`

Reply:
Done, I renamed it to `create` and simplified logic as proposed